eng/Subsets.props

```
@@ -49,6 +49,11 @@
     <_NativeAotSupportedOS Condition="'$(TargetOS)' == 'windows' or '$(TargetOS)' == 'linux' or '$(TargetOS)' == 'osx' or '$(TargetOS)' == 'maccatalyst' or '$(TargetOS)' == 'iossimulator' or '$(TargetOS)' == 'ios' or '$(TargetOS)' == 'tvossimulator' or '$(TargetOS)' == 'tvos' or '$(TargetOS)' == 'freebsd'">true</_NativeAotSupportedOS>
     <_NativeAotSupportedArch Condition="'$(TargetArchitecture)' == 'x64' or '$(TargetArchitecture)' == 'arm64' or '$(TargetArchitecture)' == 'arm' or '$(TargetArchitecture)' == 'loongarch64' or '$(TargetArchitecture)' == 'riscv64' or ('$(TargetOS)' == 'windows' and '$(TargetArchitecture)' == 'x86')">true</_NativeAotSupportedArch>
     <NativeAotSupported Condition="'$(_NativeAotSupportedOS)' == 'true' and '$(_NativeAotSupportedArch)' == 'true'">true</NativeAotSupported>
+
+    <!-- Determine if AOT tools can run on the specified target -->
+    <_AotToolsSupportedOS Condition="'$(TargetsMobile)' != 'true' and '$(TargetsLinuxBionic)' != 'true' and '$(StageOneBuild)' != 'true'">true</_AotToolsSupportedOS>
+    <_AotToolsSupportedArch>true</_AotToolsSupportedArch>
+    <AotToolsSupported Condition="'$(_AotToolsSupportedOS)' == 'true' and '$(_AotToolsSupportedArch)' == 'true'">true</AotToolsSupported>
```

```suggestion
    <AotToolsSupported Condition="'$(NativeAotSupported)' == 'true' and '$(TargetsMobile)' != 'true' and '$(TargetsLinuxBionic)' != 'true' and '$(StageOneBuild)' != 'true'">true</AotToolsSupported>
```
Delete this invariant?

---

I prefer keeping it as separate os/arch properties to match the above patterns.

Also adding `NativeAotSupported` to the condition is technically incorrect, as we can build crossgen2 and ilc to run on a platform that doesn't support NativeAOT.

---

> we can build crossgen2 and ilc to run on a platform that doesn't support NativeAOT.

Is it theoretical or do we have an actual platform like that? We have `NativeAotSupported` and `UseNativeAotForComponents`, so far those two properties were enough.  I understand the rationale here (ilc.exe on host irrespective of ilc runtime pack), but if we can avoid introducing third property, that would be good.

---

I believe that case actually is the `UseNativeAotForComponents == false` case. I'll try to coalesce these properties.

---

Actually, I think AotToolsSupported is slightly broader. Basically `AotToolsSupported` can be `true` when `UseNativeAotForComponents` is `false`, as you can still run crossgen2 or ILC in single-file mode.

---

Sounds good. Once https://github.com/dotnet/source-build/issues/4955 is flowed in (with updated SDK in runtime), we can try to enable `UseNativeAotForComponents` for riscv64. We can do it today and I tried it, but apparently it has no effect, ilc in riscv64 nupkg was still having singlefilehost mode (`_IsCommunityCrossArchitecture` part of the condition is effectively a no-op in current main), hopefully we will find the explanation.

---

> Basically AotToolsSupported can be true when UseNativeAotForComponents is false, as you can still run crossgen2 or ILC in single-file mode.

* `NativeAotSupported`: whether to build AOT tools for target
* `UseNativeAotForComponents`: whether to use AOT tools during the build; it uses host ilc.exe and target runtime-pack.
* AotToolsSupported: it's like `UseNativeAotForComponents` but with singlefile mode twist to get more coverage? It still needs target runtime pack, no? The only _benefit_ is that for the host which itself does not support AOT can build target AOT due to singlefile ilc.exe. Do we have an example of such less-capable host building for more-capable target?

I think we either should not need this property or we should not care for the "less coverage" behavior of `UseNativeAotForComponents`. i.e. whatever coverage `AotToolsSupported` is trying to enhance can be done in `UseNativeAotForComponents` without introducing additional confusion to the mix. It can always use host ilc.exe, then it depends on the availability of target ilc runtime pack.

cc @jkotas, @MichalStrehovsky

---

> Basically AotToolsSupported can be true when UseNativeAotForComponents is false, as you can still run crossgen2 or ILC in single-file mode.

* `NativeAotSupported`: whether to build AOT tools for target.
* `UseNativeAotForComponents`: whether to use AOT tools during the build; it uses host ilc.exe and target runtime-pack.
* `AotToolsSupported`: it's like `UseNativeAotForComponents` but with singlefile mode twist to get more coverage? It still needs target runtime pack, no? The only _benefit_ is that for the host which itself does not support AOT can build target AOT due to singlefile ilc.exe. Do we have an example of such less-capable host building for more-capable target?

I think we either should not need this property or we should not care for the "less coverage" behavior of `UseNativeAotForComponents`. i.e. whatever coverage `AotToolsSupported` is trying to enhance can be done in `UseNativeAotForComponents` without introducing additional confusion to the mix. It can always use host ilc.exe, then it depends on the availability of target ilc runtime pack.

cc @jkotas, @MichalStrehovsky

---

`NativeAOTSupported` means "can I run NativeAOTd code on the target platform".

`AotToolsSupported` means "can I run crossgen2 or ilc on the target platform".

`UseNativeAotForComponents` means "Can the SDK in the global.json run ILC to produce an app for the target platform".

---

> AotToolsSupported means "can I run crossgen2 or ilc on the target platform".

Would it be better if we dropped the AOT detail and broadened it to mean whether .NET SDK exists on the target platform?

---

That's a good option. I'll change it to that on Monday.